File: app/src/main/java/com/platdmit/simplecloudmanager/app/fragments/LoginFragment.java
package com.platdmit.simplecloudmanager.app.fragments;

import android.content.Context;
import android.os.Bundle;
import android.os.Vibrator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.google.android.material.snackbar.Snackbar;
import com.jakewharton.rxbinding4.widget.RxTextView;
import com.platdmit.simplecloudmanager.R;
import com.platdmit.simplecloudmanager.app.helpers.UiVisibilityStatus;
import com.platdmit.simplecloudmanager.app.vm.LoginViewModel;
import com.platdmit.simplecloudmanager.app.vm.factory.LoginViewModelFactory;
import com.platdmit.simplecloudmanager.data.api.implement.ApiAccountRepoImp;
import com.platdmit.simplecloudmanager.domain.SCMApp;
import com.platdmit.simplecloudmanager.domain.converters.implement.AccountConvertImp;
import com.platdmit.simplecloudmanager.domain.repo.AccountRepo;
import com.platdmit.simplecloudmanager.domain.repo.implement.AccountRepoImp;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

import static androidx.navigation.Navigation.findNavController;

public class LoginFragment extends Fragment {
    private static final String TAG = LoginFragment.class.getSimpleName();
    private LoginViewModel mLoginViewModel;
    private InputMethodManager mInputMethodManager;
    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();

    private ConstraintLayout mLoginFormLayout;
    private ConstraintLayout mPinFormLayout;
    private ConstraintLayout mLoaderLayout;

    private EditText mAccountLogin;
    private EditText mAccountPass;
    private EditText mAccountPinCode;
    private Button mAccountSubmit;
    private Button mOnDemoAccount;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        //Hide Bottom menu and Toolbar
        setUiVisibleStatus(false);

        if(savedInstanceState != null){
            mLoginViewModel = new ViewModelProvider(this).get(LoginViewModel.class);
        } else {
            mLoginViewModel = new ViewModelProvider(this,
                    new LoginViewModelFactory<AccountRepo>(
                            new AccountRepoImp(
                                    new ApiAccountRepoImp(),
                                    SCMApp.getDb(),
                                    new AccountConvertImp()
                            ), AccountRepo.class, SCMApp.getActualApiKeyService()
                    )).get(LoginViewModel.class);
        }

        View v = inflater.inflate(R.layout.fragment_login, container, false);
        mInputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);

        mLoginFormLayout = v.findViewById(R.id.login_layout);
        mPinFormLayout = v.findViewById(R.id.pin_layout);
        mLoaderLayout = v.findViewById(R.id.loader_layout);

        mAccountLogin = v.findViewById(R.id.user_login);
        mAccountPass = v.findViewById(R.id.user_pass);
        mAccountPinCode = v.findViewById(R.id.user_pin_code);

        mOnDemoAccount = v.findViewById(R.id.form_demo_submit);
        mOnDemoAccount.setOnClickListener(view -> {
            mLoginViewModel.onDemoAccount();
        });

        mAccountSubmit = v.findViewById(R.id.form_submit);
        mAccountSubmit.setOnClickListener(view -> {
            view.setEnabled(false);
            mLoginViewModel.addNewAccount(mAccountLogin.getText().toString(), mAccountPass.getText().toString());
        });

        mLoginViewModel.getAuthStatus().observe(getViewLifecycleOwner(), this::authStatusHandler);
        mLoginViewModel.getRegStatus().observe(getViewLifecycleOwner(), this::regStatusHandler);

        return v;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mCompositeDisposable.clear();
        setUiVisibleStatus(true);
    }

    private void regStatusHandler(LoginViewModel.LoginFormStatus status){
        switch (status){
            case AUTH_INVALID:
                showTextAlert(R.string.login_submit_not_found);
                break;
            case NEED_SET_PIN:
                showTextAlert(R.string.login_submit_correct);
                mLoginFormLayout.setVisibility(View.GONE);
                pinFormInit(true);
                break;
        }
    }

    private void authStatusHandler(LoginViewModel.LoginFormStatus status){
        switch (status){
            case YES_ACTIVE_USER:
                mLoaderLayout.setVisibility(View.GONE);
                pinFormInit(false);
                break;
            case NOT_ACTIVE_USER:
                mLoaderLayout.setVisibility(View.GONE);
                loginFormInit();
                break;
            case PIN_INVALID:
                authFall();
                break;
            case ON_DEMO:
                authDemo();
                break;
            case SUCCESS:
                authSuccess();
                break;
        }
    }

    private void pinFormInit(boolean isNew){
        if(isNew){
            TextView formTitle = getView().findViewById(R.id.pin_layout_title);
            formTitle.setText(R.string.pin_layout_title_new);
            mCompositeDisposable.add(
                    RxTextView.textChanges(mAccountPinCode)
                            .filter(pinText -> pinText.toString().length() == 4)
                            .map(CharSequence::toString)
                            .subscribe(mLoginViewModel::addNewAccountPin)
            );
        } else {
            mCompositeDisposable.add(
                    RxTextView.textChanges(mAccountPinCode)
                            .filter(pinText -> pinText.toString().length() == 4)
                            .map(CharSequence::toString)
                            .subscribe(mLoginViewModel::checkAccountPin)
            );
        }
        mPinFormLayout.setVisibility(View.VISIBLE);
        mAccountPinCode.requestFocus();
        mInputMethodManager.showSoftInput(mAccountPinCode, InputMethodManager.SHOW_IMPLICIT);
    }

    private void loginFormInit(){
        mLoginFormLayout.setVisibility(View.VISIBLE);

        mCompositeDisposable.add(Observable.combineLatest(
                RxTextView.textChanges(mAccountLogin),
                RxTextView.textChanges(mAccountPass),
                (login, pass) -> login.length() > 0 && pass.length() > 0)
                .subscribe(mAccountSubmit::setEnabled)
        );
    }

    private void authSuccess(){
        try {
            mInputMethodManager.hideSoftInputFromWindow(getView().getWindowToken(), 0);
            findNavController(getView()).popBackStack();
            findNavController(getView()).navigate(R.id.serverListFragment);
        } catch (NullPointerException ignored) {
        }
    }

    private void authFall(){
        showTextAlert(R.string.login_pin_incorrect);
        showVibratorAlert();
        mAccountPinCode.getText().clear();
    }

    private void authDemo() {
        try {
            mInputMethodManager.hideSoftInputFromWindow(getView().getWindowToken(), 0);
            this.getViewModelStore().clear();
            findNavController(getView()).navigate(R.id.serverListFragment);
            setUiVisibleStatus(true);
        } catch (NullPointerException ignored) {
        }
    }

    private void setUiVisibleStatus(boolean status){
        try {
            ((UiVisibilityStatus)getActivity()).setVisibilityToolbar(status);
            ((UiVisibilityStatus)getActivity()).setVisibilityNavigation(status);
        } catch (NullPointerException ignored){
        }
    }

    private void showTextAlert(int stringId){
        Snackbar.make(getView(), stringId, Snackbar.LENGTH_SHORT).show();
    }

    private void showVibratorAlert(){
        try {
            Vibrator vibrator = (Vibrator) getContext().getSystemService(Context.VIBRATOR_SERVICE);
            if (vibrator.hasVibrator()) {
                vibrator.vibrate(200);
            }
        } catch (NullPointerException ignored){
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/fragments/LoginFragment.kt
package com.platdmit.simplecloudmanager.fragments

import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.core.content.getSystemService
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.Navigation
import com.google.android.material.snackbar.Snackbar
import com.jakewharton.rxbinding4.widget.textChanges
import com.platdmit.simplecloudmanager.R
import com.platdmit.simplecloudmanager.helpers.UiVisibilityStatus
import com.platdmit.simplecloudmanager.states.LoginState
import com.platdmit.simplecloudmanager.vm.LoginViewModel
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.functions.BiFunction
import kotlinx.android.synthetic.main.fragment_login.*

@AndroidEntryPoint
class LoginFragment : Fragment(R.layout.fragment_login) {
    private val loginViewModel: LoginViewModel by viewModels()
    private val inputMethodManager: InputMethodManager? = context?.getSystemService()
    private val compositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //Hide Bottom menu and Toolbar
        setUiVisibleStatus(false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        loginViewModel.loginStateLiveData.observe(viewLifecycleOwner, Observer { stateHandler(it) })

        //On demo mode
        form_demo_submit.setOnClickListener {
            setStateInstance(
                    LoginViewModel.StateInstance.OnDemoMode
            )
        }

        //On check auth
        form_submit.setOnClickListener {
            it.isEnabled = false
            setStateInstance(
                    LoginViewModel.StateInstance.NewAccount(user_login.text.toString(), user_pass.text.toString())
            )
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
        setUiVisibleStatus(true)
    }

    private fun stateHandler(loginState: LoginState){
        when(loginState){
            is LoginState.ActiveUserYes -> {
                loader_layout.visibility = View.GONE
                pinFormInit(false)
            }
            is LoginState.ActiveUserNo -> {
                loader_layout.visibility = View.GONE
                loginFormInit()
            }
            is LoginState.UserNeedPin -> {
                showTextAlert(R.string.login_submit_correct)
                login_layout.visibility = View.GONE
                pinFormInit(true)
            }
            is LoginState.PinInvalid -> {
                authFall()
            }
            is LoginState.AuthInvalid -> {
                showTextAlert(R.string.login_submit_not_found)
                showVibratorAlert()
            }
            is LoginState.OnDemo -> {
                authDemo()
            }
            is LoginState.Success -> {
                authSuccess()
            }
            is LoginState.Loading -> {
                authDemo()
            }
        }
    }

    private fun setStateInstance(stateInstance: LoginViewModel.StateInstance){
        loginViewModel.setStateInstance(stateInstance)
    }

    private fun pinFormInit(isNew: Boolean) {
        if (isNew) {
            pin_layout_title.setText(R.string.pin_layout_title_new)
            compositeDisposable.add(
                    user_pin_code.textChanges()
                            .filter { it.toString().length == 4 }
                            .map { it.toString() }
                            .subscribe {
                                setStateInstance(
                                        LoginViewModel.StateInstance.NewAccountPin(it)
                                )
                            }
            )
        } else {
            compositeDisposable.add(
                    user_pin_code.textChanges()
                            .filter { it.toString().length == 4 }
                            .map { it.toString() }
                            .subscribe {
                                setStateInstance(
                                        LoginViewModel.StateInstance.CheckAccountPin(it)
                                )
                            }
            )
        }
        pin_layout.visibility = View.VISIBLE
        user_pin_code.requestFocus()
        inputMethodManager?.showSoftInput(user_pin_code, InputMethodManager.SHOW_IMPLICIT)
    }

    private fun loginFormInit() {
        login_layout.visibility = View.VISIBLE
        compositeDisposable.add(Observable.combineLatest(
                user_login.textChanges(),
                user_pass.textChanges(),
                BiFunction { login: CharSequence, pass: CharSequence -> login.isNotEmpty() && pass.isNotEmpty() })
                .subscribe { form_submit.isEnabled = it }
        )
    }

    private fun authSuccess() {
        try {
            inputMethodManager?.hideSoftInputFromWindow(requireView().windowToken, 0)
            Navigation.findNavController(requireView()).popBackStack()
            Navigation.findNavController(requireView()).navigate(R.id.serverListFragment)
        } catch (ignored: NullPointerException) {
        }
    }

    private fun authFall() {
        showTextAlert(R.string.login_pin_incorrect)
        showVibratorAlert()
        user_pin_code.text.clear()
    }

    private fun authDemo() {
        try {
            inputMethodManager?.hideSoftInputFromWindow(requireView().windowToken, 0)
            this.viewModelStore.clear()
            Navigation.findNavController(requireView()).navigate(R.id.serverListFragment)
            setUiVisibleStatus(true)
        } catch (ignored: NullPointerException) {
        }
    }

    private fun setUiVisibleStatus(status: Boolean) {
        try {
            (activity as? UiVisibilityStatus)?.setVisibilityToolbar(status)
            (activity as? UiVisibilityStatus)?.setVisibilityNavigation(status)
        } catch (ignored: NullPointerException) {
        }
    }

    private fun showTextAlert(stringId: Int) {
        Snackbar.make(requireView(), stringId, Snackbar.LENGTH_SHORT).show()
    }

    private fun showVibratorAlert() {
        try {
            val vibrator : Vibrator? = context?.getSystemService()
            vibrator?.let {
                if (vibrator.hasVibrator()) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        vibrator.vibrate(VibrationEffect.createOneShot(200, VibrationEffect.DEFAULT_AMPLITUDE))
                    } else {
                        vibrator.vibrate(200)
                    }
                }
            }
        } catch (ignored: NullPointerException) {
        }
    }
}